pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,263 @@ package traceroute
+	// Ensure the message is an ICMP Time Exceeded message
+	if msg.Type != ipv4.ICMPTypeTimeExceeded {
+		return 0, nil, errors.New("message is not 'Time Exceeded'")
+	}
 
-func newTraceroute(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error) {
-	opts := &traceroute.TracerouteOptions{}
-	opts.SetTimeoutMs(timeout)
-	opts.SetRetries(retries)
-	opts.SetMaxHops(maxHops)
-	opts.SetPort(port)
-	return traceroute.Traceroute(dest, opts)
-}
+	// The first 20 bytes of Data are the IP header, so the TCP segment starts at byte 20
+	tcpSegment := msg.Body.(*icmp.TimeExceeded).Data[20:]
 
-type result struct {
-	// The minimum number of hops required to reach the target
-	NumHops int
-	// The path taken to the destination
-	Hops []hop
-}
+	// Extract the source port from the TCP segment
+	destPort := int(tcpSegment[0])<<8 + int(tcpSegment[1])
 
-type hop struct {
-	Addr    string
-	Latency time.Duration
-	Success bool
+	return destPort, routerAddr, nil
 }
 
-// Run runs the check in a loop sending results to the provided channel
-func (tr *Traceroute) Run(ctx context.Context, cResult chan checks.ResultDTO) error {
-	ctx, cancel := logger.NewContextWithLogger(ctx)
-	defer cancel()
-	log := logger.FromContext(ctx)
+// TraceRoute performs a traceroute to the specified host using TCP and listens for ICMP Time Exceeded messages using ICMP.
+func TraceRoute(ctx context.Context, cfg tracerouteConfig) (map[int][]Hop, error) {
+	// this could also be a 2d array, but I feel like using an int map here makes the json easier to understand
+	// as it explicitly shows a mapping of ttl->hops
+	hops := make(map[int][]Hop)
+	log := logger.FromContext(ctx).With("target", cfg.Dest)
 
-	log.Info("Starting traceroute check", "interval", tr.config.Interval.String())
-	for {
-		select {
-		case <-ctx.Done():
-			log.Error("Context canceled", "error", ctx.Err())
-			return ctx.Err()
-		case <-tr.DoneChan:
-			return nil
-		case <-time.After(tr.config.Interval):
-			res := tr.check(ctx)
-
-			cResult <- checks.ResultDTO{
-				Name: tr.Name(),
-				Result: &checks.Result{
-					Data:      res,
-					Timestamp: time.Now(),
-				},
-			}
-			log.Debug("Successfully finished traceroute check run")
-		}
-	}
-}
-
-// GetConfig returns the current configuration of the check
-func (tr *Traceroute) GetConfig() checks.Runtime {
-	tr.Mu.Lock()
-	defer tr.Mu.Unlock()
-	return &tr.config
-}
-
-func (tr *Traceroute) check(ctx context.Context) map[string]result {
-	res := make(map[string]result)
-	log := logger.FromContext(ctx)
+	timeoutDuration := time.Duration(cfg.Timeout) * time.Second
 
-	type internalResult struct {
-		addr string
-		res  result
+	addr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf("%s:%d", cfg.Dest, cfg.Port))
+	if err != nil {
+		log.Error("failed to resolve target name", "err", err.Error())
+		return nil, err
 	}
 
+	results := make(chan Hop, cfg.MaxHops)
```

Comment:
Mhm, I don't like that channel construct here but I don't have a better idea so far.

But nevertheless, the channel should have a size of `cfg.MaxHops * cfg.Rc` to provide an appropriate size in case of retrying at least one hop and max out the maxHops.